main/src/main/scala/sbt/internal/server/BuildServerReporter.scala

```diff
@@ -90,22 +90,32 @@ final class BuildServerReporterImpl(
     if (ref.id().contains("<")) None
     else Some(converter.toPath(ref))
 
+  /**
+   * Send diagnostics from the compilation to the client.
+   * Do not send empty diagnostics if previous ones were also empty ones.
+   *
+   * @param analysis current compile analysis
+   */
   override def sendSuccessReport(
       analysis: CompileAnalysis,
-      prev: CompileAnalysis,
-      reportAllPreviousProblems: Boolean
   ): Unit = {
-    val prevInfos = prev.readSourceInfos().getAllSourceInfos().asScala
+    val shouldReportAllProblems = !bspCompileState.compiledAtLeastOnce.getAndSet(true)
     for {
       (source, infos) <- analysis.readSourceInfos.getAllSourceInfos.asScala
       filePath <- toSafePath(source)
     } {
-      val prevProblems = prevInfos.get(source).map(_.getReportedProblems()).getOrElse(Array.empty)
-      val dontPublish = prevProblems.length == 0 && infos.getReportedProblems().length == 0
-      val shouldPublish = reportAllPreviousProblems || !dontPublish
+      // clear problems for current file
+      val hadProblems = bspCompileState.hasAnyProblems.remove(filePath)
+      val reportedProblems = infos.getReportedProblems.toVector
+
+      // publish diagnostics if:
+      // 1. file had any problems previously - we might want to update them with new ones
+      // 2. file has fresh problems - we might want to update old ones
+      // 3. build project is compiled first time - shouldReportAllProblems is set
+      val shouldPublish = hadProblems || reportedProblems.nonEmpty || shouldReportAllProblems
 
       if (shouldPublish) {
-        val diagnostics = infos.getReportedProblems.toSeq.flatMap(toDiagnostic)
+        val diagnostics = reportedProblems.flatMap(toDiagnostic)
```

Comment:
Here diagnostics can contain some warnings. We need to set `hasAnyProblems` if it is not empty otherwise the warnings would never disappear.
![warning](https://user-images.githubusercontent.com/13123162/174976780-2ef67833-db2c-4404-9c4e-f598d4031112.gif)

```suggestion
        val diagnostics = reportedProblems.flatMap(toDiagnostic)
        if (diagnostics.nonEmpty) bspCompileState.hasAnyProblems.add(filePath)
```